fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTestValidator.java

```diff
@@ -0,0 +1,239 @@
+import com.apple.foundationdb.record.lucene.directory.FDBDirectoryManager;
+import com.apple.foundationdb.record.lucene.search.LuceneOptimizedIndexSearcher;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.expressions.GroupingKeyExpression;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
+import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
+import com.apple.foundationdb.record.query.expressions.Comparisons;
+import com.apple.foundationdb.record.query.plan.ScanComparisons;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.base.Verify;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.TopDocs;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Assertions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * A utility for validating the consistency and contents of a lucene index.
+ */
+public class LuceneIndexTestValidator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneIndexTestValidator.class);
+    private final Supplier<FDBRecordContext> contextProvider;
+    private final Function<FDBRecordContext, FDBRecordStore> schemaSetup;
+
+    public LuceneIndexTestValidator(Supplier<FDBRecordContext> contextProvider, Function<FDBRecordContext, FDBRecordStore> schemaSetup) {
+        this.contextProvider = contextProvider;
+        this.schemaSetup = schemaSetup;
+    }
+
+    /**
+     * A broad validation of the lucene index, asserting consistency, and that various operations did what they were
+     * supposed to do.
+     * <p>
+     *     This has a lot of validation that could be added, and it would be good to be able to control whether it's
+     *     expected that `mergeIndex` had been run or not; right now it assumes it has been run.
+     * </p>
+     * @param index the index to validate
+     * @param expectedIds a map from group to primaryKey to timestamp
```

Comment:
Is there any reason for this (and the subroutines it calls) to use `id` instead of `primaryKey`?